### PR TITLE
Stream manager performance improvement

### PR DESCRIPTION
**Pull request by congwang**
Major changes include:
1. Add memory pool for dedicated protobuf messages to reuse object, reducing the expensive cost to new/delete a protobuf message.
2. Reduce duplicated protobuf message, which was design to be immutable and be easy for reasoning. But it is super inefficient.
3. Optimize *.proto to reduce duplicated serialization/deserialization cost. For instance, when one stmgr receives a message from another stmgr, it does not need to parse the TupleSet message; it just need to send its serialized byte array to corresponding instance.
4. Reduce duplicated computation. For instance, protobuf Message.ByteSize(), which is expensive and re-calculate the result (not reading from cache value computed) every time invoked.
5. Use a C++11 feature to optimize protobuf performance.

6X performance improvement for clang (Tested on mac); 4x performance improvement for gcc on Linux (Tested on Centos7 for large scale).

Joint work with Maosong Fu.

### CODE REVIEW

File: heron/common/src/cpp/network/server.h [hunk 1]
```diff
@@ -200,6 +208,19 @@ class Server : public BaseServer {
   // Called when the connection is closed
   virtual void HandleConnectionClose_Base(BaseConnection* connection, NetworkErrorCode _status);
 
+  // TODO(mfu):
```
Comment (kramasamy) [inline heron/common/src/cpp/network/server.h]: What is this TODO? Can you comment on this?

Comment (maosongfu) [inline heron/common/src/cpp/network/server.h]: Removed.

File: heron/common/tests/cpp/network/switch_unittest.cpp [hunk 2]
```diff
@@ -49,7 +49,7 @@ class Terminate : public Client {
 
   virtual void HandleConnect(NetworkErrorCode _status) {
     if (_status == OK) {
-      TerminateMessage* message = new TerminateMessage();
+      TerminateMessage message;
```
Comment (kramasamy) [inline heron/common/tests/cpp/network/switch_unittest.cpp]: SendMessage might be deleting memory underneath - we need to be careful here.

Comment (kramasamy) [inline heron/common/tests/cpp/network/switch_unittest.cpp]: ignore this.

File: heron/stmgr/src/cpp/manager/stmgr.h [hunk 3]
```diff
@@ -150,6 +152,16 @@ class StMgr {
   sp_string zkroot_;
   sp_int32 metricsmgr_port_;
   sp_int32 shell_port_;
+
+  // TODO(mfu):
```
Comment (kramasamy) [inline heron/stmgr/src/cpp/manager/stmgr.h]: Please remove the TODO - since there is no comment what needs to be done.

**Comment by kramasamy** [comment]
@congwang and @maosongfu - CI is failing.

File: heron/stmgr/src/cpp/manager/stmgr.h (hunk 3 above)
Comment (congwang) [inline heron/stmgr/src/cpp/manager/stmgr.h]: Removed.

**Comment by kramasamy** [comment]
👍

**Comment by objmagic** [comment]
Can you briefly describe how you improved the performance? Can we have some numbers that shows the performance difference?

**Comment by maosongfu** [comment]
@objmagic Added detailed descriptions for major changes and performance improvement.

**Comment by kramasamy** [comment]
@congwang @maosongfu - looks like all the tests passed, should we go ahead and merge it?

File: heron/stmgr/src/cpp/manager/stmgr.cpp [hunk 4]
```diff
@@ -120,6 +121,12 @@ void StMgr::Init() {
   CHECK_GT(eventLoop_->registerTimer([this](EventLoop::Status status) {
     this->UpdateProcessMetrics(status);
   }, true, PROCESS_METRICS_FREQUENCY), 0);
+
+  //
```
Comment (congwang) [inline heron/stmgr/src/cpp/manager/stmgr.cpp]: This needs to be removed.

File: heron/stmgr/src/cpp/manager/stmgr.h [hunk 5]
```diff
@@ -150,6 +152,15 @@ class StMgr {
   sp_string zkroot_;
   sp_int32 metricsmgr_port_;
   sp_int32 shell_port_;
+
+  proto::system::HeronTupleSet2 current_control_tuple_set_;
+  std::vector<sp_int32> out_tasks;
```
Comment (congwang) [inline heron/stmgr/src/cpp/manager/stmgr.h]: This should be named as "out_tasks_".

File: heron/stmgr/src/cpp/manager/stream-consumers.h [hunk 6]
```diff
@@ -37,7 +40,15 @@ class StreamConsumers {
   void NewConsumer(const proto::api::InputStream& _is, const proto::api::StreamSchema& _schema,
                    const std::vector<sp_int32>& _task_ids);
 
-  void GetListToSend(const proto::system::HeronDataTuple& _tuple, std::list<sp_int32>& _return);
+  void GetListToSend(const proto::system::HeronDataTuple& _tuple, std::vector<sp_int32>& _return);
+
+  inline bool isShuffleGrouping() {
```
Comment (congwang) [inline heron/stmgr/src/cpp/manager/stream-consumers.h]: No need 'inline' explicitly, inline is default here.

File: heron/stmgr/src/cpp/util/tuple-cache.h [hunk 7]
```diff
@@ -58,18 +64,41 @@ class TupleCache {
...
                             sp_uint64* _tuples_cache_max_tuple_size);
     void add_ack_tuple(const proto::system::AckTuple& _tuple, sp_uint64* total_size_);
     void add_fail_tuple(const proto::system::AckTuple& _tuple, sp_uint64* total_size_);
     void add_emit_tuple(const proto::system::AckTuple& _tuple, sp_uint64* total_size_);
 
     void drain(sp_int32 _task_id,
-               std::function<void(sp_int32, proto::system::HeronTupleSet*)> _drainer);
+               std::function<void(sp_int32, proto::system::HeronTupleSet2*)> _drainer);
+
+     // TODO(mfu):
```
Comment (congwang) [inline heron/stmgr/src/cpp/util/tuple-cache.h]: Should be removed.

File: heron/stmgr/src/cpp/util/tuple-cache.h [hunk 8]
```diff
@@ -58,18 +64,41 @@ class TupleCache {
...
+    std::vector<proto::system::HeronTupleSet2*> _heron_tuple_set_pool;
+
+    inline proto::system::HeronTupleSet2* acquire() {
```
Comment (congwang) [inline heron/stmgr/src/cpp/util/tuple-cache.h]: No 'inline'

**Comment by kramasamy** [comment]
@congwang - is this completed?

**Comment by maosongfu** [comment]
This looks good to me. :+1:

**Pull request merged**